fix: siwe-parcel-compatibility

Harman-singh-waraich opened the pull request:
## Context
Siwe fails to build when using with parcel bundler. 

If using `ethers v5`,
Parcel expects` import {verifyMessage} from "ethers"` to resolve, and fails to build

<img width="823" alt="Screenshot 2024-04-11 at 3 36 18 PM" src="https://github.com/spruceid/siwe/assets/51452848/bae576c7-9f76-4394-8571-aea01c042401">

If using `ethers v6`,
Parcel expects `import {utils} from "ethers" `to resolve and fails to build

<img width="819" alt="Screenshot 2024-04-11 at 3 39 32 PM" src="https://github.com/spruceid/siwe/assets/51452848/84b825be-a27d-4273-8550-5f6b08178674">

## Minimal Reproduction 
https://codesandbox.io/p/devbox/shy-morning-jvjffk-forked-jw8vcj

Steps:
- yarn 
- yarn build

## Solution 
Use try catch to handle the error and export the functions [verifyMessage, hashMessage, getAddress] based on ethers version
All tests passing ✅ 

```
let ethersVerifyMessage = null;
let ethersHashMessage = null;
let ethersGetAddress = null;

try {
  const { utils } = require('ethers');
  ethersVerifyMessage = utils.verifyMessage;
  ethersHashMessage = utils.hashMessage;
  ethersGetAddress = utils.getAddress;
} catch (error) {
  const { verifyMessage, getAddress, hashMessage } = require('ethers');

  ethersVerifyMessage = verifyMessage as (
    message: Uint8Array | string,
    sig: Ethers6SignatureLike
  ) => string;

  ethersHashMessage = hashMessage as (message: Uint8Array | string) => string;

  ethersGetAddress = getAddress as (address: string) => string;
}

export const verifyMessage = ethersVerifyMessage;
export const hashMessage = ethersHashMessage;
export const getAddress = ethersGetAddress;
```

sbihel reviewed (changes requested):
This looks good. Apart from the lint errors (you can run `npn run lint` to run the linter), could you add a check in the catch to see if the `ethers` peer dependency is installed?

Harman-singh-waraich commented:
> This looks good. Apart from the lint errors (you can run `npn run lint` to run the linter), could you add a check in the catch to see if the `ethers` peer dependency is installed?

- Updated the fix to use es6 imports. ( passes the linter and tests ✅ )
- The code should throw an error itself if `ethers` is not installed 🤔 , can you elaborate more on how we can check if `ethers` is installed. Also since we import `ethers` at top itself, the bundler should notify the user if `ethers` is not resolved

sbihel commented:
> The code should throw an error itself if ethers is not installed 🤔 , can you elaborate more on how we can check if ethers is installed. Also since we import ethers at top itself, the bundler should notify the user if ethers is not resolved

That's fair, it was just in case it wouldn't be clear enough.

sbihel commented:
Actually, just one more thing, does this PR address #187?

Harman-singh-waraich commented:
> Actually, just one more thing, does this PR address #187?

It does not but i see it's somewhat related, in ethers v6 , we use the `Providers` from `ethers.Providers` whereas in  ethers v5 it's imported from `ethers.providers.Provider`.

So, in addition to `verifyMessage`, `getAddress` and `hashMessage`, this could also be added in the `ethersCompact` file , to check for v6 and v5

sbihel commented:
This matches my understanding, would you mind doing it?

Harman-singh-waraich commented:
> This matches my understanding, would you mind doing it?

Added a check for providers, and moved the logic to ethersCompact.
I cannot reproduce the same error as in #187 , but i think this should resolve it. 🤔

sbihel commented:
Brilliant, thank you. Released as `siwe@2.3.0`.
